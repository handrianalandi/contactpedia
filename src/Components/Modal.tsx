import styled from "@emotion/styled";
import React, { ForwardedRef, forwardRef } from "react";
import { Button, Modal as ModalElement } from "react-bootstrap";

const ModalStyle = styled(ModalElement)`
  @keyframes slideInFromBottom {
    0% {
      transform: translateY(100%);
    }
    100% {
      transform: translateY(0);
    }
  }

  & .modal-dialog {
    position: fixed;
    top: auto;
    right: auto;
    left: auto;
    bottom: 0;
    width: 100%;
    margin: 0;
    max-height: 90vh;

    & .modal-body {
      max-height: 70vh;
      overflow-y: auto;
    }
  }

  & .modal-content {
    border-bottom-left-radius: 0;
    border-bottom-right-radius: 0;
    animation: 0.3s ease-out 0s 1 slideInFromBottom;
  }

  & .modal-footer {
    border-top: none;
    & .btn-primary {
      background-color: #06ba63;
      border: 1px solid #06ba63;
      font-weight: 500;
    }
  }

  & .modal-header {
    border-bottom: none;
  }

  @media (min-width: 768px) {
    & .modal-dialog {
      position: relative;
      display: flex;
      align-items: center;
      margin-right: auto;
      margin-left: auto;
      height: 100%;
    }

    & .modal-content {
      border-radius: var(--bs-modal-border-radius);
    }
  }
`;

interface ModalProps {
  isOpen: boolean;
  heading?: string;
  confirmText?: string;
  confirmButtonType?: "primary" | "danger";
  onClose: () => void;
  onConfirm?: () => void;
  children?: React.ReactNode;
}

export default function Modal({
  isOpen,
  heading,
  confirmText = "Confirm",
  confirmButtonType = "primary",
  onClose,
  onConfirm,
  children,
}: ModalProps) {
  return (
    <ModalStyle show={isOpen} onHide={onClose}>
      {heading && (
        <ModalElement.Header closeButton>
          <ModalElement.Title>{heading}</ModalElement.Title>
        </ModalElement.Header>
      )}
      <ModalElement.Body>
        {children}
    </ModalElement.Body>
      <ModalElement.Footer>
        <Button variant="outline-secondary" onClick={onClose}>
          Close
        </Button>
        <Button className="confirmBtn" variant={confirmButtonType} onClick={onConfirm}>
          {confirmText}
        </Button>
      </ModalElement.Footer>
    </ModalStyle>
  );
}