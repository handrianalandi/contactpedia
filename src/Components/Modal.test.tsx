import {
    render,
    screen,
    fireEvent} from '@testing-library/react';
import Modal from './Modal';

describe('Modal', () => {
    const props = {
        isOpen: true,
        heading: 'Test',
        confirmText: 'Test Confirm',
        onClose: jest.fn(),
        onConfirm: jest.fn(),
    };

    it('should render correctly opened', () => {
        render(<Modal isOpen/>);
        const modal = screen.getByTestId('modal-test');
        expect(modal).toMatchSnapshot();
    }
    );

    it('should render correctly when isLoading', () => {
        render(<Modal isOpen isLoading/>);
        const modal = screen.getByTestId('modal-test');
        expect(modal).toMatchSnapshot();
    }
    );

    it('should render correctly with props and children', () => {
        render(
        <Modal {...props}>
            <div>Test</div>
        </Modal>
        );
        const modal = screen.getByTestId('modal-test');
        expect(modal).toMatchSnapshot();
    }
    );

    it('should call onClose when close button is clicked', () => {
        render(<Modal {...props}/>);
        const closeButton = screen.getByLabelText('Close');
        fireEvent.click(closeButton);
        expect(props.onClose).toHaveBeenCalled();
    }
    );

    it('should call onConfirm when confirm button is clicked', () => {
        render(<Modal {...props}/>);
        const confirmButton = screen.getByTestId('modal-confirm-button');
        fireEvent.click(confirmButton);
        expect(props.onConfirm).toHaveBeenCalled();
    }
    );


});