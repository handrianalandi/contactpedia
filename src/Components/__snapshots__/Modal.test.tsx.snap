// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render correctly opened 1`] = `
<div
  class="modal-dialog"
  data-testid="modal-test"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-body"
    />
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-primary"
        data-testid="modal-confirm-button"
        type="button"
      >
        Confirm
      </button>
    </div>
  </div>
</div>
`;

exports[`Modal should render correctly when isLoading 1`] = `
<div
  class="modal-dialog"
  data-testid="modal-test"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-body"
    />
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-primary"
        data-testid="modal-confirm-button"
        disabled=""
        type="button"
      >
        <div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-spinner fa-spin fa-sm  css-1vbona3"
            data-icon="spinner"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M304 48a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zm0 416a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zM48 304a48 48 0 1 0 0-96 48 48 0 1 0 0 96zm464-48a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zM142.9 437A48 48 0 1 0 75 369.1 48 48 0 1 0 142.9 437zm0-294.2A48 48 0 1 0 75 75a48 48 0 1 0 67.9 67.9zM369.1 437A48 48 0 1 0 437 369.1 48 48 0 1 0 369.1 437z"
              fill="currentColor"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Modal should render correctly with props and children 1`] = `
<div
  class="modal-dialog"
  data-testid="modal-test"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Test
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <div>
        Test
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-primary"
        data-testid="modal-confirm-button"
        type="button"
      >
        Test Confirm
      </button>
    </div>
  </div>
</div>
`;
